Add support for python.org builds targeting osx 10.9

robbuckley opened the pull request:
as mentioned in https://github.com/MacPython/pandas-wheels/issues/37, I've extended the osx parts of multibuild to allow building wheels using 10.9-based pythons from [python.org](https://www.python.org/), as well as the default 10.6. I've built wheels on travis for `pandas` using my fork, and they seem to install OK on my local system (10.14) with a range of pythons  (2.7.10 / maxosx-10.14-intel, 3.7.2 / macosx-10.6-intel, 3.6.7 / maxosc-10.7-x86_64, 3.7.1 / macosx-10.9-x86_64)

My implementation adds a new environment variable `MB_PYTHON_OSX_VER`, to be defined in the end user's `.travis.yml`. It can determines the  mac osx version of the python downloaded from python.org to build the wheel, for example:
```
python-2.7.15-macosx10.6.pkg
python-2.7.15-macosx10.9.pkg
```
If not present, it defaults to 10.6, as before. Wheels built using 10.9-based [64b] python are tagged with `macosx_10_9_x86_64.macosx_10_10_x86_64` platforms, those built with 10.6-based [32/64b] pythons are tagged as before with `macosx_10_6_intel.macosx_10_9_intel.macosx_10_9_x86_64.macosx_10_10_intel.macosx_10_10_x86_64`

I have come across a few issues which i cant explain. I don't know how important they are:
1. `tests/test_supported_wheels.sh` doesn't seem to be able to fail - i discovered this when I broke the wheel format during testing.
2. recent versions of pip I've tried (18.1, 19.0.1) seem to flag both the 10.6-based [tagged as 32/64b] and 10.9-based [tagged as 64b] wheels I've built as compatible with my [64b] system with both single arch [64b] and dual-arch [32/64b] pythons. That seems to contradict the info on https://github.com/MacPython/wiki/wiki/Spinning-wheels, which states that pip will not install single-arch wheels on dual-arch python versions. I have  more info if needed
3. using `otool`, i can see that libs in pandas wheels that I've built using the python.org 10.9-based pythons are fat 32/64b. I was expecting the libs to be 64b only. The wheel when first built is  tagged as i expected with `macosx_10_9_x86_64`, it just that the binaries seem to be been compiled as dual-arch
4. I have followed the existing code, adding 10.10 tags to the 10.9 wheels. Im curious why 10.10, and not also other later macosx versions (10.11...). I suppose this was to workaround some issue on older versions of pip?

radarhere commented on osx_utils.sh:
```diff
@@ -298,12 +306,20 @@ function repair_wheelhouse {
     # Add platform tags to label wheels as compatible with OSX 10.9 and
     # 10.10.  The wheels will be built against Python.org Python, and so will
-    # in fact be compatible with OSX >= 10.6.  pip < 6.0 doesn't realize
+    # in fact be compatible with OSX >= 10.6 or >=10.9. pip < 6.0 doesn't realize
```
In isolation, I feel that this comment is now confusing. >= 10.6 **or** >= 10.9? Perhaps add a note that this depends on `MB_PYTHON_OSX_VER`?

robbuckley replied:
clarified this coment

native-api commented on tests/test_multibuild.sh:
```diff
@@ -8,7 +8,9 @@ source tests/test_common_utils.sh
     source osx_utils.sh
-    get_macpython_environment $PYTHON_VERSION $VENV
+    # osx_utils defines MACPYTHON_DEFAULT_OSX
+    MB_PYTHON_OSX_VER=${MB_PYTHON_OSX_VER:-$MACPYTHON_DEFAULT_OSX}
```
Can we avoid the hack and logic duplication with `travis_osx_steps.sh`? This should be done by stock logic like with other envvars.
AFAICS it's safe to move this to `osx_utils.sh` and get rid of `MACPYTHON_DEFAULT_OSX` altogether.

robbuckley replied:
Agree its a bit of a hack, but i wanted to allow `MB_PYTHON_OSX_VER` to be optional in the end users `.travis.yml` file, with a default value set in a single place. Since the function `get_macpython_environment` already has an optional parameter `venv`, and is called directly from both `test_multibuild.sh` and `travis_osx_steps.sh`, this seemed the least disruptive solution.  I'm not that experienced with bash, so maybe there's a better solution, but its not obvious to me.

robbuckley replied:
OK so i have managed to remove the duplicated logic by substituting
```
get_macpython_environment $PYTHON_VERSION $MB_PYTHON_OSX_VER $VENV`
```
with
```
get_macpython_environment $PYTHON_VERSION ${MB_PYTHON_OSX_VER:-""} $VENV 
```
in each case

native-api commented on tests/test_osx_utils.sh:
```diff
@@ -12,9 +12,32 @@
 [ "$(pyinst_ext_for_version 3)" == pkg ] || ingest
 [ "$(pyinst_fname_for_version 2.7.8)" == "python-2.7.8-macosx10.6.dmg" ] || ingest
 [ "$(pyinst_fname_for_version 2.7.9)" == "python-2.7.9-macosx10.6.pkg" ] || ingest
+[ "$(pyinst_fname_for_version 2.7.10)" == "python-2.7.10-macosx10.6.pkg" ] || ingest
```
Just one case for the extra argument will cover all new logic. All other cases are redundant.

robbuckley replied:
i added all the 'missing' versions after 2.7.9 / 3.6.1 which are available on [python.org](https://www.python.org/ftp/python/), as well as the versions available for 10.9. 

I'll cut this list right down again. Its probably worth replacing the existing versions that are tested (2.7.8/9, 3.5.0, 3.6.1) with something newer. as the are all rather old now

native-api commented on tests/test_supported_wheels.sh:
```diff
@@ -10,6 +10,15 @@ lpv=$(lex_ver $PYTHON_VERSION)
     for whl in wheel==0.31.1 wheel==0.32.0 wheel; do
         $pip_install -U $whl
-        $PYTHON_EXE supported_wheels.py tornado-5.1-cp27-cp27m-macosx_10_6_intel.whl tornado-5.1-cp36-cp36m-macosx_10_13_x86_64.whl texext-0.6.1-cp36-none-any.whl
+        $PYTHON_EXE supported_wheels.py \
```
There are no changes to supported_wheels.py, nor to its spec. Looks redundant.

robbuckley replied:
not related to this PR i agree, but maybe worth doing in a separate PR? 

Given the intent of `test_supported_wheels.sh` / `supported_wheels.py` seems to be try to pip install all the wheel versions which the pep425 tags say are applicable to the current python version (repeated for each python version in the CI build matrix). 

Let me know and i will push another change to remove

native-api replied:
`suported_wheels.py` merely filters out "invalid" names from its stdin

robbuckley replied:
i can see that, but doesnt `supported_wheels.sh` try to install the (filtered) list of wheels? I dont know if you can install >1 wheel at a time on pip...

native-api commented on travis_osx_steps.sh:
```diff
@@ -5,11 +5,12 @@ set -e
 # These load common_utils.sh
 source $MULTIBUILD_DIR/osx_utils.sh
+# osx_utils defines MACPYTHON_DEFAULT_OSX
+MB_PYTHON_OSX_VER=${$MB_PYTHON_OSX_VER:-$MACPYTHON_DEFAULT_OSX}
```
duplicate logic, see above

robbuckley replied:
removed, see above

native-api commented:
> I have come across a few issues which i cant explain. I dont know how important they are:
> 
>     1. `tests/test_supported_wheels.sh` doesn't seem to be able to fail - i discovered this when I broke the wheel format during testing.

Yes, the test is broken. It runs the script but doesn't check its result.

>     2. recent versions of pip I've tried (18.1, 19.0.1) seem to flag both the 10.6-based [tagged as 32/64b] and 10.9-based [tagged as 64b] wheels I've built as compatible with my [64b] system with both single arch [64b] and dual-arch [32/64b] pythons. That seems to contradict the info on https://github.com/MacPython/wiki/wiki/Spinning-wheels, which states that pip will not install single-arch wheels on dual-arch python versions. I have  more info if needed

If this is true, it looks like `pip`'s problem.
 
>     3. using `otool`, i can see that libs in pandas wheels that I've built using the python.org 10.9-based pythons are fat 32/64b. I was expecting the libs to be 64b only. The wheel when first built is  tagged as i expected with `macosx_10_9_x86_64`, it just that the binaries seem to be been compiled as dual-arch

See https://github.com/matthew-brett/multibuild/blob/devel/configure_build.sh#L19 .

robbuckley commented on .gitignore:
```diff
@@ -1,2 +1,3 @@
 downloads/
 *.orig
+.idea/
```
that must have sneaked in - its PyCharm related. I will remove it

robbuckley commented on README.rst:
```diff
@@ -250,6 +255,22 @@ To use these scripts
+          env:
+            - MB_PYTHON_VERSION=3.6
+            - MB_PYTHON_OSX_VER=10.9
+        # Default OSX (xcode image) is 10.13 (xcode 9.4.1) as of 2018-08-03
```
OK, agree should not be in the readme.

robbuckley replied:
done

robbuckley commented on .travis.yml:
```diff
@@ -108,8 +122,21 @@ matrix:
         - VENV=venv
-    # Default OSX (xcode image) is 10.13 (xcode 9.4.1) as of 2018-08-03
-    # See: https://docs.travis-ci.com/user/reference/osx/
+#    # Default OSX (xcode image) is 10.13 (xcode 9.4.1) as of 2018-08-03
```
its not originally my comment, see https://github.com/matthew-brett/multibuild/blob/851e08b39ab9c458be9f8fd4b54900c08b1d6730/.travis.yml#L111

native-api replied:
Then why are they in the diff in the first place? There must be some kind of change here.

Oh, I see. You added a hash at the start. Which is redundant still :-)

robbuckley replied:
oops. I will fix this at the end, i dont want to restart travis builds for this

robbuckley replied:
fixed

robbuckley commented on osx_utils.sh:
```diff
@@ -258,22 +265,23 @@ function get_macpython_environment {
     #     $venv_dir : {directory_name|not defined}
     #         If defined - make virtualenv in this directory, set python / pip
     #         commands accordingly
+    #     $osx_version: {10.6 | 10.9}
```
done

robbuckley commented on osx_utils.sh:
```diff
@@ -144,13 +148,16 @@ function install_macpython {
+    #   $py_version
+    #       Version given in major or major.minor or major.minor.micro e.g
+    #       "3" or "3.4" or "3.4.1".
+    #   $osx_version = { 10.6 | 10.9 }
```
done

robbuckley commented on osx_utils.sh:
```diff
@@ -130,11 +132,13 @@ function install_macpython {
     # Parameters:
     #     $version : [implementation-]major[.minor[.patch]]
     #         The Python implementation to install, e.g. "3.6" or "pypy-5.4"
+    #     $osx_version: { 10.6|10.9 }. Ignored for pypy
```
done both (renamed and added explanatory comment)

robbuckley commented on osx_utils.sh:
```diff
@@ -119,9 +120,10 @@ function pyinst_fname_for_version {
     # echo filename for OSX installer file given Python version
     # Parameters
     #   $py_version (python version in major.minor.extra format)
+    #   $osx_ver    (macosx version in major.minor format, defaults to MACPYTHON_DEFAULT_OSX)
```
done

robbuckley commented on README.rst:
```diff
@@ -238,6 +239,10 @@ To use these scripts
+        - os: osx
+          env:
+            - MB_PYTHON_VERSION=2.7
+            - MB_PYTHON_OSX_VER=10.9
```
done

robbuckley commented:
please note, i added another commit to build for 64-bit only for 10.9-based pythons, with a new function `mac_arch_for_pyosx_version` in `osx_utils.sh`. Once its passing the multibuild CI I will repeat the checks 2 and 3 in my first comment locally, and also build + test pandas wheels using a fork of https://github.com/MacPython/pandas-wheels

robbuckley commented:
OK, i have built `pandas` 0.24.0 wheels for 2.7, 3.6 and 3.7, both 10.6 (dual arch) and 10.9 (single arch versions). They all passed the `pandas` release tests.

Also i did the following tests on my local system
1. pip installed
- 3.6 wheels on anaconda 3.6.7 python built for `macosx-10.7-x86_64`
- 3.7 wheels on conda forge 3.7.1 python built for `macosx-10.9-x86_64`
- 3.7 wheels on `python.org` installed on my base system (3.7.2 built for `macosx-10.6-intel`)
Results: both 10.6(dual arch) and 10.9(single arch) wheels installed. This was expected on the first two pythons which are 64b-only. But not on the 3rd which is dual-arch. According to X that should have refused to install the 64b-only wheel 

2. checked the c++ standard library dependencies for a `pandas` C extension using `delocate-listdeps`
- 10.6 builds link to `libstdc++`
- 10.9 builds link to `libc++`
(both as expected)

3. checked the arches in a sample `pandas` C extension `.so` using `lipo -info`
- 10.6 (dual arch) builds - `i386 x86_x64`
- 10.9 (64b only) builds - `x86_64`
(as expected)

Thats all the testing I was planning to do for this PR

matthew-brett commented:
@radarhere - I'm slammed at the moment - can I leave you to merge this if you are happy?

radarhere commented on README.rst:
```diff
@@ -56,10 +57,17 @@ functions and variables in earlier scripts:
+OS X builds may be targetted either at macOS 10.6+
+(dual arch 64 / 32 bit) or macOS 10.9+ (64b only). These depend on the
+corresponding build of python from https://www.python.org/downloads/mac-osx/.
+At the time of writing, 10.9+ / 64 bit build are supported for python
```
```suggestion
At the time of writing, 10.9+ / 64 bit builds are supported for python
```

radarhere commented on README.rst:
```diff
@@ -56,10 +57,17 @@ functions and variables in earlier scripts:
 
 See ``travis_osx_steps.sh`` to review source order.
 
-The macOS build / test and phase are on the macOS VM started by Travis CI.
```
Very minor: Why a comma?

radarhere commented on osx_utils.sh:
```diff
@@ -255,25 +295,30 @@ function get_macpython_environment {
-    local venv_dir=$2
-
+    local py_osx_ver=$2
+    local venv_dir=$3
```
This isn't backwards compatible. Could you keep `venv_dir` as the second argument, rather than break existing code?

robbuckley replied:
should be no pb to swap the argument order. My thinking is it feels more logical to keep the arguments relating to the python version together. 

Since this function isnt suppsed to be called by end user code / `.travis.yml`, are you concerned about future merge conflicts?

yet another approach is to have use an environment variable - eg just have the lower level code  access `MB_PYTHON_OSX_VER` directly, but this will be messy to test properly (global state) and seems to go against the style of the existing codebase

radarhere replied:
I have found https://github.com/scikit-image/scikit-image/blob/master/tools/travis/osx_install.sh#L16

native-api replied:
First, whoever uses internal APIs accepts the risk of breakage. The README mentions a few functions -- those can be considered public ones.

Second, `multibuild`'s docs (i.e. the README) don't make any formal backward compatibility promises whatsoever. So it can be retained all other things being equal, but not at the cost of technical debt.

matthew-brett replied:
Honestly I haven't made any real attempt to delineate public APIs.

So my rule is - break back compatibility if there's a really good reason to.

Here, I don't think that's true - yes it would be nicer to reorganize the arguments but it's not essential.

robbuckley replied:
I've swapped the order of arguments `$2` and `$3` of `get_macpython_environment` to preserve back compatibility. Also I have `osx_utils.sh` use `MB_PYTHON_OSX_VER` as a default so that the function can be called without passing it on the 3rd argument, which avoids the ugliness of forcing the caller to pass "" as the 2nd arg when no `venv` is needed. Hope this addresses all the angles.

radarhere commented on osx_utils.sh:
```diff
@@ -116,25 +117,58 @@ function pyinst_ext_for_version {
+        echo "Invalid python osx version: ${py_osx_ver}, supported values: 10.6 and 10.9"
+        exit 1
+    fi
+    echo
```
Why echo a blank line?

robbuckley replied:
whoops! its gone

radarhere commented on tests/test_osx_utils.sh:
```diff
@@ -10,11 +10,14 @@
-[ "$(pyinst_fname_for_version 2.7.8)" == "python-2.7.8-macosx10.6.dmg" ] || ingest
-[ "$(pyinst_fname_for_version 2.7.9)" == "python-2.7.9-macosx10.6.pkg" ] || ingest
-[ "$(pyinst_fname_for_version 3.4.1)" == "python-3.4.1-macosx10.6.dmg" ] || ingest
-[ "$(pyinst_fname_for_version 3.4.2)" == "python-3.4.2-macosx10.6.pkg" ] || ingest
```
I presume you're just updating these numbers to be more recent. However, they're not arbitrary. There is a difference between [2.7.8 and 2.7.9](https://github.com/matthew-brett/multibuild/blob/c96230626ceae3108c4d2f4e3081f989514028e4/osx_utils.sh#L104) and between [3.4.1 and 3.4.2](https://github.com/matthew-brett/multibuild/blob/c96230626ceae3108c4d2f4e3081f989514028e4/osx_utils.sh#L110)

robbuckley replied:
Yes, i was just updating the `pyinst_fname_for_version` tests to more recent Python versions. The difference in functionality between 3.7.8/9 and 3.4.1/2 is tested by lines 4 + 5 and 8 + 9 where `pyinst_ext_for_version` behaviour is checked, it seems redundant to check here as well.

robbuckley commented:
I've rerun the tests listed in https://github.com/matthew-brett/multibuild/pull/224#issuecomment-463392679, all OK. That covers building `cpython` wheels, but not `pypy`. Since almost all of these changes are `cpython` specific, the risk should hopefully be low that some issue has been introduced there.

radarhere commented on osx_utils.sh:
```diff
@@ -116,25 +118,77 @@ function pyinst_ext_for_version {
+}
+
+function get_macpython_osx_ver {
+    # get minimum macOS version (e.g. 10.9) from the disutils platform tag
```
```suggestion
    # get minimum macOS version (e.g. 10.9) from the distutils platform tag
```

radarhere commented on osx_utils.sh:
```diff
@@ -116,25 +118,77 @@ function pyinst_ext_for_version {
+}
+
+function get_macpython_arch {
+    # get arch (e.g. intel or x86_64) from the disutils platform tag
```
```suggestion
    # get arch (e.g. intel or x86_64) from the distutils platform tag
```

radarhere commented on tests/test_python_install.sh:
```diff
@@ -58,3 +60,18 @@ else # not virtualenv
     fi
 fi
+
+# for cpython, check macos version and arch are as expected
```
```suggestion
# for CPython, check macOS version and arch are as expected
```

radarhere commented on tests/test_python_install.sh:
```diff
@@ -58,3 +60,18 @@ else # not virtualenv
+distutils_plat=$($PYTHON_EXE -c "import distutils.util; print(distutils.util.get_platform())")
+echo "Python cmd archs: $(lipo -info $(which $PYTHON_EXE))"
+if [[ $requested_impl = 'cp' ]]; then
+    echo "Cpython, checking platform..."
```
```suggestion
    echo "CPython, checking platform..."
```

radarhere commented on tests/test_python_install.sh:
```diff
@@ -58,3 +60,18 @@ else # not virtualenv
+        ingest "Wrong Python platform tag: ${distutils_plat}!=${expected_tag}"
+    fi
+elif [[ $requested_impl = 'py' ]]; then
+    echo "Pypy, skipping platform check..."
```
```suggestion
    echo "PyPy, skipping platform check..."
```

radarhere commented on tests/test_osx_utils.sh:
```diff
@@ -10,17 +10,28 @@
-    ingest
+[ "$(get_py_mm_nodot)" == $(echo "${cpython_version:0:3}" | tr -d .) ] || ingest
+
+# test lookup of arch from cpython macos target build
```
```suggestion
# test lookup of arch from CPython macOS target build
```

robbuckley commented:
i've addressed all comments (i think). There's another issue i've noticed, and have a fix for: the arch flags for pypy wheels will be set according to the `MB_PYTHON_OSX_VER` variable, which is only meant to be for cpython. I have a fix prepared, shall I add it to this PR, or wait till its merged and submit as a new PR? The effect of not adding this is fairly benign, in that pypy wheels may be built for dual arch when thy should only be single arch. But the way its linked to the environment variable may be confusing. @radarhere  @matthew-brett

radarhere commented on osx_utils.sh:
```diff
@@ -116,25 +118,91 @@ function pyinst_ext_for_version {
+    #                       the function get_distutils_platform, provided by
+    #                       common_utils.sh. Fails if this is not a mac platform
+    #
+    # Note: MUST only be called after the version of python used to build the
```
```suggestion
    # Note: MUST only be called after the version of Python used to build the
```

radarhere commented on osx_utils.sh:
```diff
@@ -116,25 +118,91 @@ function pyinst_ext_for_version {
+    #                       the function get_distutils_platform, provided by
+    #                       common_utils.sh. Fails if this is not a mac platform
+    #
+    # Note: MUST only be called after the version of python used to build the
```
```suggestion
    # Note: MUST only be called after the version of Python used to build the
```

radarhere commented:
I would add it to this PR

radarhere commented on tests/test_osx_utils.sh:
```diff
@@ -10,17 +10,35 @@
+[ "$(macpython_arch_for_version 2.7 10.9)" == "x86_64" ] || ingest
+[ "$(macpython_arch_for_version pypy-2.7)" == "x86_64" ] || ingest
+
+# test lookup of arch / min macOS versions from installed python distutils tag
```
```suggestion
# test lookup of arch / min macOS versions from installed Python distutils tag
```

robbuckley commented:
i've build `pandas` 0.24 wheels again with the latest version of this PR, and checked them per https://github.com/matthew-brett/multibuild/pull/224#issuecomment-463392679. As a final check, I installed one of the 10.9 64b-only wheels in a conda environment and ran the `pandas` test suite.

robbuckley commented:
is there anything more you'd like me to do on this, @radarhere , @matthew-brett ? Im quite keen to get this merged and released as I'd like to use it in an MR for the next `pandas` release. If i can help pls let me know.

matthew-brett commented:
@radarhere - I'm happy when you're happy.

matthew-brett commented:
Great - thanks both for all your work on this.

robbuckley commented:
thanks both for bearing with me on this PR! Does it need a release (i suppose that means a merge of devel>master) before I can use it for production?

radarhere commented:
You don't need devel to be merged into master.

https://github.com/matthew-brett/multibuild/blob/devel/README.rst
> We try to keep the master branch stable and do testing and development in the devel branch. From time to time we merge devel into master.
> 
> In practice, you can check out the newest commit from devel that works for you, then stay at it until you need newer features.

If you're talking about the use of multibuild in https://github.com/MacPython/pandas-wheels, then provided they don't mind going with a commit from the devel branch instead of master, you just need to [go into the submodule, check out the relevant commit, go back into the base repository, and create a PR out of the change](https://stackoverflow.com/a/10916398/4093019). To potentially save you some time though, here's a PR and/or commit that you can use if you want - https://github.com/MacPython/pandas-wheels/compare/master...radarhere:cpython-10.9?expand=1

robbuckley commented:
thanks

robbuckley commented:
> Hmm - yes - it's complicated. Thinking about it, for macOS, we have:

> - The architecture of the installed Python. At the moment this gets set implicitly by the choice of 10.6 (intel) or 10.9 (x86_64)
> - The architecture to compile for. This is getting set by this code, so forced to the same as the architecture of installed Python.

Yes, youre right. Similar considerations apply to the targetted minimum macOS version set by`MACOSX_DEPLOYMENT_TARGET`. I took the path of least resistance, and just installed the python version with the right values for both.

> I guess we can use PLAT as the second thing - architecture to compile for, and default to it being the same as the Python architecture. This will result in a lot of x86_64 compiles for intel - because people don't notice the distinction, and have PLAT set in their travis files. I doubt that will cause a problem. What do you think?

That would be consistent with `PLAT` usage for linux, and makes the "user API" more consistent. But its not back compatible - until now setting `PLAT` in the users `.travis.yml` had no effect for macOS. I cant judge which is better... 

If you do change this, then really the wheel tags should be fixed to be consistent with the arches built. That could be done by reinstating some logic based on `$PLAT` in `repair_wheelhouse`

BTW i can see my changes broke things for some other users (eg https://github.com/matthew-brett/multibuild/pull/242). I'd like to help, e.g. by adding test coverage for breakages, even if they are now fixed. So if there's anything else you can point me towards, please let me know.

matthew-brett commented:
It turned out your changes causes trouble only because of the PLAT issue - i.e. I thought PLAT was the PLAT set in `.travis.yml`, but in fact it was the one forced from `configure_build.sh`. 

The most common use-case for PLAT differing from the implied platform, from the Python installation, is building `intel` wheels which do in fact only contain `x86_64`.  I think that's fine - it allows the wheels to install on `intel` builds, and only breaks if you try to use it in 32-bit mode, which, as you say, is unlikely to happen.

Any interest in dropping some documentation of PLAT etc into the README, though?

robbuckley commented:
Yes, Im happy to add to the docs. 

So do you want to add the ability for the user to override `PLAT` for macOS builds or are you OK with it the way it is?

pls confirm and I'll open an issue to remind me to update the docs. 

The problem with conversing on a closed PR is that the only notification from github is by email when a new commet is added - it doesnt appear on my dashboard at all. So Im always afraid I'll miss something
